### PR TITLE
architecture, part II

### LINKED ISSUE #158: Rust mdBook
Intros:

- [x] add most important links (ideally also our API/code documentation)
- [x] it should be clear where which code should be added

Arc42:

- [x] integrate all current markdown files (architecture, usecases, etc.)
- [x]  quality goals:
        1. stability
        2. maintenance
        3. performance
        4. privacy
- [x] solution strategy
       - description of 3 layers
       - types/structs
       - diesel/rust intro in docs(Gabriel)
       - be stateless (all state in DB)

### PR DESCRIPTION
**Pull request by markus2330**
fixes #158

## Basics

- [ ] Details of what you changed are in commit messages.
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.
- [ ] The PR is rebased with current master.

## Checklist

- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added unit tests for my code
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

### CODE REVIEW

File: backend/src/controller/plantings.rs [hunk 1]
```diff
@@ -104,7 +104,7 @@ pub async fn create(new_seed_json: Json<NewPlantingDto>, pool: Data<Pool>) -> Re
         ("oauth2" = [])
     )
 )]
-#[patch("/{id}")]
+#[put("/{id}")]
```
Comment (Bushuo) [inline backend/src/controller/plantings.rs:107]: Why not PATCH?
The actions should be as fine-grained as possible, e.g. the move action should only change x and y of a planting. 
Under the hood, it calls the update endpoint with patch, sending just
```json
{ "x": 123, "y": 123 }
```
These are the semantics of PATCH as I understand it.

PUT means that the whole dto would have to be sent. Making it impossible to undo/redo fine-grained changes.

File: doc/guidelines/logging.md [hunk 2]
```diff
@@ -2,42 +2,51 @@
...
+
+To log you can simply call the macros in the `log` crate.
+
+For example to log using the debug level:
+
+```rust,ignore
+log::debug!("This is my variable: {}", some_variable);
+```
+
+As you can see the behavior is the same as it is for example for `println!()`.
```
Comment (Bushuo) [inline doc/guidelines/logging.md:48]:
```suggestion
As you can see, it can be used in the same way as for example `println!()`.
```

**Review by Bushuo — CHANGES_REQUESTED** [review]
I agree with everything except that we always use PUT.

File: backend/src/controller/plantings.rs (hunk 1 above)
Comment (markus2330) [reply backend/src/controller/plantings.rs:107]: If we use the idempotent subset of PATCH it is fine, too. The way you describe it, it is perfectly okay. And I agree, when the DTOs become bigger (like Plant) we definitely want to only send minimal updates.

Do we already know how the actions in the state are described? Will they be like the PATCH example you describe here?

Comment (Bushuo) [reply backend/src/controller/plantings.rs]: > Do we already know how the actions in the state are described? Will they be like the PATCH example you describe here?
>
Yes, they look quite similar.
Though the types can not be shared.
An action in the frontend also encapsulates the behavior: 
- how to apply itself to the state.
- how to construct its inverse action.
- how to communicate itself to the backend.

**Review by Bushuo — APPROVED** [review]

File: doc/guidelines/api.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+## Endpoints
+
+The endpoint paths use:
+
+- hierarchical structure
+- nouns instead of verbs
+- only plural (exception: config)
+- all endpoints need authorization (exception: config)
+- all paths below `/api`
+- layer-specific paths below `/layers/<name of layer>/` e.g. `/layers/plants/plantings`
```
Comment (Bushuo) [inline doc/guidelines/api.md:24]:
```suggestion
- layer-specific paths below `maps/<map_id>/layers/<name of layer>/` e.g. `/layers/plants/plantings`
```
We will need the map_id as part of every request to the layer.
Should this be encoded in the path or as part of the request body?

File: backend/src/controller/plantings.rs (hunk 1 above)
Comment (markus2330) [reply backend/src/controller/plantings.rs]: Yes, I see that there is more information in the state but what about having the shared DTO as sub-object?

**Pull request merged**